### PR TITLE
feat(ARCO-180): reorg e2e tests

### PR DESCRIPTION
**Pull request by kuba-4chain**
## Description of Changes

Provide a brief description of the changes you've made.

## Linked Issues / Tickets

Reference any related issues or tickets, e.g. "Closes #123".

## Testing Procedure

Describe the tests you've added or any testing steps you've taken.

- [ ] I have added new unit tests
- [ ] All tests pass locally
- [ ] I have tested manually in my local environment

## Checklist:

- [ ] I have performed a self-review of my own code
- [ ] I have made corresponding changes to the documentation
- [ ] My changes generate no new warnings
- [ ] I have updated `CHANGELOG.md` with my changes

### CODE REVIEW

File: internal/blocktx/processor.go [hunk 1]
```diff
@@ -662,55 +720,221 @@ func (p *Processor) markTransactionsAsMined(ctx context.Context, blockID uint64,
...
+	}
+
+	err = p.store.UpdateBlocksStatuses(ctx, blockStatusUpdates)
+	if err != nil {
+		return nil, err
+	}
+
+	p.logger.Info("reorg performed successfully")
+
+	registeredTxs, err := p.store.GetRegisteredTxsByBlockHashes(ctx, append(staleHashes, longestHashes...))
```
Comment (arkadiuszos4chain) [inline internal/blocktx/processor.go:823]: With two requests you will simplfy logic

File: internal/blocktx/processor.go [hunk 2]
```diff
@@ -662,55 +720,221 @@ func (p *Processor) markTransactionsAsMined(ctx context.Context, blockID uint64,
...
+		case blocktx_api.Status_STALE:
+			staleTxs = append(staleTxs, tx)
+		default:
+			// do nothing - ignore txs from ORPHANED or UNKNOWN blocks
+		}
+	}
+
+	staleTxs = findDistinctStaleTxs(longestTxs, staleTxs)
+
+	return append(longestTxs, staleTxs...), nil
```
Comment (arkadiuszos4chain) [inline internal/blocktx/processor.go:844]: You can return multidimensional slice here and gain on memory usage

File: internal/blocktx/processor_test.go [hunk 3]
```diff
@@ -250,14 +255,22 @@ func TestHandleBlock(t *testing.T) {
 	}
 }
 
-func TestHandleBlockReorg(t *testing.T) {
+func TestHandleBlockReorgAndOrphans(t *testing.T) {
```
Comment (arkadiuszos4chain) [inline internal/blocktx/processor_test.go:258]: Divide into two tests?

File: internal/blocktx/store/postgresql/get_orphaned_chain.go [hunk 4]
```diff
@@ -0,0 +1,86 @@
...
+		WITH RECURSIVE orphans AS (
+			SELECT
+				hash
+				,prevhash
+				,merkleroot
+				,height
+				,processed_at
+				,status
+				,chainwork
+			FROM blocktx.blocks WHERE hash = $1 AND status = $2
```
Comment (arkadiuszos4chain) [inline internal/blocktx/store/postgresql/get_orphaned_chain.go:36]: You could stop execution on first non-processed ancestor

File: internal/blocktx/store/postgresql/get_stale_chain.go [hunk 5]
```diff
@@ -47,7 +45,6 @@ func (p *PostgreSQL) GetStaleChainBackFromHash(ctx context.Context, hash []byte)
 				,b.merkleroot
 				,b.height
 				,b.processed_at
-				,b.orphanedyn
 				,b.status
 				,b.chainwork
 			FROM blocktx.blocks b JOIN prevBlocks p ON b.hash = p.prevhash AND b.status = $2
```
Comment (arkadiuszos4chain) [inline internal/blocktx/store/postgresql/get_stale_chain.go:50]: You could stop execution on first non-processed ancestor

File: internal/blocktx/processor.go [hunk 6]
```diff
@@ -426,206 +449,241 @@ func (p *Processor) processBlock(msg *p2p.BlockMessage) (err error) {
...
 	blockHash := msg.Header.BlockHash()
-	previousBlockHash := msg.Header.PrevBlock
-	merkleRoot := msg.Header.MerkleRoot
+	blockHeight := msg.Height
 
-	// don't process block that was already processed
+	p.logger.Info("processing incoming block", slog.String("hash", blockHash.String()), slog.Uint64("height", blockHeight))
+
+	// check if we've already processed that block
 	existingBlock, _ := p.store.GetBlock(ctx, &blockHash)
```
Comment (arkadiuszos4chain) [inline internal/blocktx/processor.go:458]: I think we shouldn't return non-processed blocks by default from any repository method.
IMO situation when we need non-processed block is exceptional and should be another method for this.

Notice that this change will simplify logic in few places

File: test/submit_05_reorg_test.go [hunk 7]
```diff
@@ -0,0 +1,147 @@
...
+	// get new UTXO for tx2
+	node_client.SendToAddress(t, bitcoind, address, float64(0.002))
+	utxos = node_client.GetUtxos(t, bitcoind, address)
+	require.True(t, len(utxos) > 0, "No UTXOs available for the address")
+
+	// make sure to pick the correct UTXO
+	var utxo node_client.UnspentOutput
+	for _, u := range utxos {
+		if u.Amount == float64(0.002) {
+			utxo = u
```
Comment (boecklim) [inline test/submit_05_reorg_test.go:48]: I think you could also find this utxo by the ID

```go
	txID := node_client.SendToAddress(t, bitcoind, address, float64(0.002))
	utxos = node_client.GetUtxos(t, bitcoind, address)
	require.True(t, len(utxos) > 0, "No UTXOs available for the address")

	// make sure to pick the correct UTXO
	var utxo node_client.UnspentOutput
	for _, u := range utxos {
		if u.Txid == txID {
			utxo = u
		}
	}
```

File: test/submit_05_reorg_test.go [hunk 8]
```diff
@@ -0,0 +1,147 @@
...
+	// prepare txStale
+	node_client.SendToAddress(t, bitcoind, address, float64(0.003))
+
+	utxos = node_client.GetUtxos(t, bitcoind, address)
+	require.True(t, len(utxos) > 0, "No UTXOs available for the address")
+
+	// make sure to pick the correct UTXO
+	for _, u := range utxos {
+		if u.Amount == float64(0.003) {
+			utxo = u
```
Comment (boecklim) [inline test/submit_05_reorg_test.go:82]: I think you could also find this utxo by the ID
```go
	txID := node_client.SendToAddress(t, bitcoind, address, float64(0.002))
	utxos = node_client.GetUtxos(t, bitcoind, address)
	require.True(t, len(utxos) > 0, "No UTXOs available for the address")

	// make sure to pick the correct UTXO
	var utxo node_client.UnspentOutput
	for _, u := range utxos {
		if u.Txid == txID {
			utxo = u
		}
	}
```

File: test/submit_05_reorg_test.go [hunk 9]
```diff
@@ -0,0 +1,147 @@
...
+	statusResp = getRequest[TransactionResponse](t, statusURL)
+	require.Equal(t, StatusMined, statusResp.TxStatus)
+	require.Equal(t, staleHash, *statusResp.BlockHash)
+
+	// verify that tx2 is now MINED_IN_STALE_BLOCK
+	statusURL = fmt.Sprintf("%s/%s", arcEndpointV1Tx, tx2.TxID())
+	statusResp = getRequest[TransactionResponse](t, statusURL)
+	require.Equal(t, StatusMinedInStaleBlock, statusResp.TxStatus)
+	require.Equal(t, tx2BlockHash, *statusResp.BlockHash)
+}
```
Comment (boecklim) [inline test/submit_05_reorg_test.go:140]: One additional interesting check would be to check a callback has been sent for tx2 with status `MINED_IN_STALE_BLOCK`

File: internal/blocktx/processor_helpers.go [hunk 10]
```diff
@@ -5,49 +5,45 @@ import (
...
+		sum = sum.Add(sum, chainwork)
 	}
+
+	return sum
 }
 
-func getLowestHeight(blocks []*blocktx_api.Block) uint64 {
-	if len(blocks) == 0 {
-		return 0
+func findDistinctStaleTxs(longestTxs, staleTxs []store.TransactionBlock) []store.TransactionBlock {
```
[the remark this answers is not in the document]
Comment (kuba-4chain) [reply internal/blocktx/processor_helpers.go:32]: Done, renamed to `exclusiveRightTxs`, params to `leftTxs` and `rightTxs`.

File: internal/blocktx/processor.go [hunk 11]
```diff
@@ -209,16 +211,31 @@ func (p *Processor) StartBlockProcessing() {
...
 
-				p.logger.Info("received block", slog.String("hash", blockhash.String()))
-				err := p.processBlock(blockMsg)
+				err = p.processBlock(blockMsg)
 				if err != nil {
-					p.logger.Error("block processing failed", slog.String("hash", blockhash.String()), slog.String("err", err.Error()))
-					p.unlockBlock(p.ctx, &blockhash)
+					p.logger.Error("block processing failed", slog.String("hash", blockHash.String()), slog.String("err", err.Error()))
+					p.unlockBlock(p.ctx, &blockHash)
+					p.stopBlockProcessGuard(&blockHash) // release guardian
```
[the remark this answers is not in the document]
Comment (kuba-4chain) [reply internal/blocktx/processor.go:224]: Done

File: internal/blocktx/processor.go [hunk 12]
```diff
@@ -426,206 +449,241 @@ func (p *Processor) processBlock(msg *p2p.BlockMessage) (err error) {
...
+	case blocktx_api.Status_ORPHANED:
+		txsToPublish, err = p.handleOrphans(ctx, block)
+	default:
+		return ErrUnexpectedBlockStatus
+	}
 
-	competing, err := p.competingChainsExist(ctx, incomingBlock)
 	if err != nil {
-		p.logger.Error("unable to check for competing chains", slog.String("hash", blockHash.String()), slog.Uint64("height", msg.Height), slog.String("err", err.Error()))
+		// error is already logged in each method above
```
[the remark this answers is not in the document]
Comment (kuba-4chain) [reply internal/blocktx/processor.go:484]: Done

File: internal/blocktx/processor.go [hunk 13]
```diff
@@ -426,206 +449,241 @@ func (p *Processor) processBlock(msg *p2p.BlockMessage) (err error) {
...
 		}
 
-		// find competing chains back to the common ancestor
-		// get all registered transactions
-		// prepare msg with competing blocks
-		incomingBlock.Status = blocktx_api.Status_STALE
-
-		if hasGreatestChainwork {
-			p.logger.Info("reorg detected - updating blocks", slog.String("hash", blockHash.String()), slog.Uint64("height", incomingBlock.Height))
+		p.logger.Info("publishing tx", slog.String("txHash", getHashStringNoErr(tx.TxHash)))
```
[the remark this answers is not in the document]
Comment (kuba-4chain) [reply internal/blocktx/processor.go:497]: Done

File: internal/blocktx/processor.go (hunk 1 above)
Comment (kuba-4chain) [reply internal/blocktx/processor.go:823]: Done

File: internal/blocktx/processor.go (hunk 2 above)
Comment (kuba-4chain) [reply internal/blocktx/processor.go:844]: Done, returning 2 slices

File: internal/blocktx/store/postgresql/get_orphaned_chain.go (hunk 4 above)
Comment (kuba-4chain) [reply internal/blocktx/store/postgresql/get_orphaned_chain.go:36]: Done, in most of other queries as well

File: internal/blocktx/store/postgresql/get_stale_chain.go (hunk 5 above)
Comment (kuba-4chain) [reply internal/blocktx/store/postgresql/get_stale_chain.go:50]: Done, in most of other queries as well

File: internal/blocktx/processor.go (hunk 6 above)
Comment (kuba-4chain) [reply internal/blocktx/processor.go:458]: Good point, done and removed the logic

File: internal/blocktx/processor.go [hunk 14]
```diff
@@ -426,206 +449,241 @@ func (p *Processor) processBlock(msg *p2p.BlockMessage) (err error) {
 		tracing.EndTracing(span, err)
 	}()
 
-	timeStart := time.Now()
-
 	blockHash := msg.Header.BlockHash()
-	previousBlockHash := msg.Header.PrevBlock
-	merkleRoot := msg.Header.MerkleRoot
+	blockHeight := msg.Height
```
[the remark this answers is not in the document]
Comment (kuba-4chain) [reply internal/blocktx/processor.go:453]: Done, added to `BlockMessage` in `internal/blocktx/blockchain_communication/block_message.go`

File: internal/blocktx/processor.go [hunk 15]
```diff
@@ -426,206 +449,241 @@ func (p *Processor) processBlock(msg *p2p.BlockMessage) (err error) {
...
+	ctx, span := tracing.StartTracing(ctx, "verifyAndInsertBlock", p.tracingEnabled, p.tracingAttributes...)
+	defer func() {
+		tracing.EndTracing(span, err)
+	}()
 
-	if !merkleRoot.IsEqual(calculatedMerkleTree[len(calculatedMerkleTree)-1]) {
-		p.logger.Error("merkle root mismatch", slog.String("hash", blockHash.String()))
-		return err
-	}
+	blockHash := msg.Header.BlockHash()
```
[the remark this answers is not in the document]
Comment (kuba-4chain) [reply internal/blocktx/processor.go:514]: Done

File: test/submit_05_reorg_test.go (hunk 7 above)
Comment (kuba-4chain) [reply test/submit_05_reorg_test.go:48]: Great, thanks, I was looking for a more reliable way. Changed

File: test/submit_05_reorg_test.go (hunk 8 above)
Comment (kuba-4chain) [reply test/submit_05_reorg_test.go:102]: Changed

**Review by boecklim — APPROVED** [review]

**Review by arkadiuszos4chain — APPROVED** [review]

**Pull request merged**